lib/async_event_emitter.js

```diff
@@ -99,20 +122,46 @@ module.exports = class AsyncEventEmitter {
    */
   async emit (eventName, ...args) {
     const listeners = this.listeners[eventName]
-    const indexesToRemove = []
 
-    if (!listeners) {
-      return
-    }
+    // TODO: Refactor
+    if (eventName !== CATCH_ALL_EVENT_NAME) {
+      const catchAllListeners = this.listeners[CATCH_ALL_EVENT_NAME]
 
-    await PI.forEach(listeners, async (l, i) => {
-      await l.cb(...args)
+      if (catchAllListeners) {
+        const indexesToRemove = []
 
-      if (l.type === 'once') {
-        indexesToRemove.unshift(i)
+        await PI.forEachSeries(catchAllListeners, async (l, i) => {
+          if (_includes(indexesToRemove, i)) {
+            return
+          }
+
```

Comment:
Couldnt you shit call `emit(CATCH_ALL_EVENT_NAME, ...args)` and let the event emitter handle the `once/on` logic?

Reply:
Done, nice refactor!